### DISCUSSION TITLE
Where to add constraints, defaults and validation when using a database

### DISCUSSION THREAD

**Discussion by silllli** [fastapi/fastapi#3042] [Questions] [answered]
I think this was answered in different places but I’m still very puzzled as to where to add model constraints, default values and data validation. Basically: what do I put in my (database) models and what do I put in my (Pydantic) schemas? I found different approaches in the FastAPI docs, issue answers and third-party FastAPI packages.

For example default values. I could set up a (SQLAlchemy) database model like this

```
models.py

class Item(Base):
    available = Column(Boolean)
```

and a Pydantic schema like this

```
schemas.py

class ItemBase(BaseModel):
    available: bool
```

To make sure the default value of `available` is `True`, I could change it to

```
models.py

class Item(Base):
    available = Column(Boolean, default=True)
```

**or**

```
schemas.py

class ItemBase(BaseModel):
    available: bool = True
```

Correct? I could also change both files, so

```
models.py

class Item(Base):
    available = Column(Boolean, default=True)
```

**and**

```
schemas.py

class ItemBase(BaseModel):
    available: bool = True
```

So which is the way to go? Do as much at the database layer, or as much at the Pydantic layer (so it gets rendered in the autogenerated docs etc.) or should I do it in both places since the Pydantic schemas should mirror the database model?

The same goes for instance for `nullable=True` in the database model or any other constraint, like string lengths.

Probably the answer to this is: it depends. But I wonder if there is a best practice with advantages and disadvantages that I can’t see at the moment. I would be glad for some guidance from experienced people.

**Comment by tiangolo** [comment, marked as answer]
You can probably use SQLModel and follow the tutorial for FastAPI: https://sqlmodel.tiangolo.com/tutorial/fastapi/

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
